missioncontrol/datalake/models.py

```diff
@@ -0,0 +1,287 @@
+import logging
+import six
+
+import json
+import datetime
+from itertools import chain
+from uuid import uuid4
+
+from connexion.exceptions import ProblemException
+from datalake_common import Metadata, InvalidDatalakeMetadata
+from django.apps import apps
+from django.conf import settings
+from django.contrib.contenttypes.fields import GenericForeignKey
+from django.contrib.contenttypes.models import ContentType
+from django.core.exceptions import ValidationError, ObjectDoesNotExist
+from django.db import models
+from django.db.models import Max
+from django.db.models.signals import pre_save, post_save
+from django.dispatch import receiver
+from django.utils import timezone, dateformat
+import boto3
+from pytz import UTC
+
+logger = logging.getLogger(__name__)
+logger.setLevel(logging.DEBUG)
+
+
+@receiver(pre_save)
+def pre_save_handler(sender, instance, *args, **kwargs):
+    # always validate local models before saving
+    if sender in apps.all_models['datalake'].values():
+        instance.full_clean()
+
+
+class DatalakeJSONEncoder(json.JSONEncoder):
+    """
+    JSONEncoder subclass that knows how to encode date/time, decimal types, and
+    UUIDs.
+    Based on DjangoJSONEncoder but uses ISO8601 with microsecond precision
+    instead of ECMA-262
+    """
+    def default(self, o):
+        # See "Date Time String Format" in the ECMA-262 specification.
+        if isinstance(o, datetime.datetime):
+            r = o.isoformat('T', timespec='microseconds')
+            if r.endswith('+00:00'):
+                r = r[:-6] + 'Z'
+            return r
+        elif isinstance(o, datetime.date):
+            return o.isoformat()
+        elif isinstance(o, datetime.time):
+            if is_aware(o):
+                raise ValueError("JSON can't represent timezone-aware times.")
+            r = o.isoformat()
+            return r
+        elif isinstance(o, datetime.timedelta):
+            return duration_iso_string(o)
+        elif isinstance(o, (decimal.Decimal, uuid.UUID, Promise)):
+            return str(o)
```

Comment:
str of a Promise doesn't sound right?
Also `Promise` doesn't seem to be defined anywhere?